src/controller/java/src/chip/tlv/types.kt

```diff
@@ -0,0 +1,167 @@
+private const val UNSIGNED_INT_TYPE = 0b00000100.toByte()
+private const val UTF8_STRING_TYPE = 0b00001100.toByte()
+private const val BYTE_STRING_TYPE = 0b00010000.toByte()
+private const val BOOLEAN_FALSE = 0b01000.toByte()
+private const val BOOLEAN_TRUE = 0b01001.toByte()
+private const val FLOATING_POINT_4 = 0b01010.toByte()
+private const val FLOATING_POINT_8 = 0b01011.toByte()
+private const val NULL = 0b10100.toByte()
+private const val STRUCTURE = 0b10101.toByte()
+private const val ARRAY = 0b10110.toByte()
+private const val LIST = 0b10111.toByte()
+private const val END_OF_CONTAINER = 0b11000.toByte()
+
+/**
+ * Represents the type of element for a TLV element.
+ *
+ * @property lengthSize the size, in bytes, of the length section of this element
+ * @property valueSize the size, in bytes, of this element's value section
+ */
+internal sealed class Type(val lengthSize: Short, val valueSize: Short) {
+  abstract fun encode(): Byte
+
+  internal companion object {
+    /** Returns the element type encoded by the given control byte. */
+    fun from(controlByte: Byte): Type {
+      // Integer and string types encode the length in the lower two bits. For these types,
+      // ignore the lower 2 bits for matching, and extract that size later.
+      val modifiedControlByte =
+        when (val byte = controlByte and MODIFIED_TYPE_MASK) {
+          SIGNED_INT_TYPE,
+          UNSIGNED_INT_TYPE,
+          UTF8_STRING_TYPE,
+          BYTE_STRING_TYPE -> byte
+          else -> controlByte and ELEMENT_TYPE_MASK
+        }
+
+      return when (modifiedControlByte) {
+        SIGNED_INT_TYPE -> SignedIntType(extractSize(controlByte))
+        UNSIGNED_INT_TYPE -> UnsignedIntType(extractSize(controlByte))
+        UTF8_STRING_TYPE -> Utf8StringType(extractSize(controlByte))
+        BYTE_STRING_TYPE -> ByteStringType(extractSize(controlByte))
+        BOOLEAN_FALSE -> BooleanType(false)
+        BOOLEAN_TRUE -> BooleanType(true)
+        FLOATING_POINT_4 -> FloatType()
+        FLOATING_POINT_8 -> DoubleType()
+        NULL -> NullType()
+        STRUCTURE -> StructureType()
+        ARRAY -> ArrayType()
+        LIST -> ListType()
+        END_OF_CONTAINER -> EndOfContainerType
+        else ->
+          throw IllegalStateException(
+            "Unexpected control byte ${modifiedControlByte.toBinaryString()}"
+          )
+      }
+    }
+
+    private fun Byte.toBinaryString() = Integer.toBinaryString(0xFF and this.toInt())
+
+    private fun extractSize(byte: Byte): Short {
+      // Variably-sized element types encode their length in the lower 2 bits.
+      return when (byte and 0b011) {
+        0b000.toByte() -> 1
+        0b001.toByte() -> 2
+        0b010.toByte() -> 4
+        else -> 8
+      }
+    }
+  }
+}
+
+private fun encodeSize(size: Short): Byte {
+  // Variably-sized element types encode their length in the lower 2 bits.
+  return when (size.toInt()) {
+    1 -> 0b000.toByte()
```

Comment:
is toByte conversion here needed? Tried it in play.kotlinlang.org and it seemed to not be required.